File: python/graphdb/query.py
"""Graph traversal & query API.

Provides a fluent :class:`GraphQuery` plus standalone ``bfs``, ``dfs`` and
``find_path`` helpers.
"""

from __future__ import annotations

from collections import deque
from typing import Deque, List, Optional, Set, TYPE_CHECKING

from .core import Node

if TYPE_CHECKING:  # pragma: no cover
    from .store import GraphStore


def _neighbors(
    store: "GraphStore",
    node_id: str,
    edge_label: Optional[str],
    direction: str,
) -> List[str]:
    """Return neighbour node ids for the given direction/edge label."""
    result: List[str] = []
    if direction in ("out", "both"):
        for edge in store.edges_from(node_id):
            if edge_label is None or edge.label == edge_label:
                result.append(edge.dst_id)
    if direction in ("in", "both"):
        for edge in store.edges_to(node_id):
            if edge_label is None or edge.label == edge_label:
                result.append(edge.src_id)
    return result


def bfs(
    store: "GraphStore",
    start_id: str,
    edge_label: Optional[str] = None,
    max_depth: int = 5,
    direction: str = "out",
) -> List[Node]:
    """Breadth-first traversal from ``start_id`` (excludes the start node)."""
    visited: Set[str] = {start_id}
    order: List[Node] = []
    queue: Deque = deque([(start_id, 0)])
    while queue:
        current, depth = queue.popleft()
        if depth >= max_depth:
            continue
        for nb in _neighbors(store, current, edge_label, direction):
            if nb not in visited:
                visited.add(nb)
                node = store.get_node_or_none(nb)
                if node is not None:
                    order.append(node)
                    queue.append((nb, depth + 1))
    return order


def dfs(
    store: "GraphStore",
    start_id: str,
    edge_label: Optional[str] = None,
    max_depth: int = 5,
    direction: str = "out",
) -> List[Node]:
    """Depth-first traversal from ``start_id`` (excludes the start node)."""
    visited: Set[str] = {start_id}
    order: List[Node] = []

    def _visit(node_id: str, depth: int) -> None:
        if depth >= max_depth:
            return
        for nb in _neighbors(store, node_id, edge_label, direction):
            if nb not in visited:
                visited.add(nb)
                node = store.get_node_or_none(nb)
                if node is not None:
                    order.append(node)
                    _visit(nb, depth + 1)

    _visit(start_id, 0)
    return order


def find_path(
    store: "GraphStore",
    src_id: str,
    dst_id: str,
    edge_label: Optional[str] = None,
    direction: str = "out",
) -> Optional[List[Node]]:
    """Return the shortest path (list of nodes) from ``src`` to ``dst``.

    Uses BFS. Returns ``None`` if no path exists. The returned list includes
    both endpoints.
    """
    if store.get_node_or_none(src_id) is None:
        return None
    if store.get_node_or_none(dst_id) is None:
        return None
    if src_id == dst_id:
        return [store.get_node(src_id)]

    visited: Set[str] = {src_id}
    prev: dict = {}
    queue: Deque[str] = deque([src_id])
    found = False
    while queue:
        current = queue.popleft()
        if current == dst_id:
            found = True
            break
        for nb in _neighbors(store, current, edge_label, direction):
            if nb not in visited:
                visited.add(nb)
                prev[nb] = current
                queue.append(nb)
    if not found and dst_id not in prev:
        return None

    # Reconstruct path.
    path_ids: List[str] = [dst_id]
    while path_ids[-1] != src_id:
        parent = prev.get(path_ids[-1])
        if parent is None:
            return None
        path_ids.append(parent)
    path_ids.reverse()
    return [store.get_node(i) for i in path_ids]


class GraphQuery:
    """A small fluent query/traversal builder over a :class:`GraphStore`."""

    def __init__(self, store: "GraphStore") -> None:
        self._store = store
        self._current: List[Node] = []
        self._matched = False

    def match(self, label: Optional[str] = None, **props) -> "GraphQuery":
        """Filter nodes by label and/or exact property values."""
        self._current = self._store.find_nodes(label=label, **props)
        self._matched = True
        return self

    def traverse(
        self,
        edge_label: Optional[str] = None,
        direction: str = "out",
        max_depth: int = 1,
    ) -> "GraphQuery":
        """Traverse from currently matched nodes, collecting reachable nodes."""
        if not self._matched:
            self._current = self._store.all_nodes()
            self._matched = True
        seen: Set[str] = set()
        collected: List[Node] = []
        for start in self._current:
            for node in bfs(
                self._store,
                start.id,
                edge_label=edge_label,
                max_depth=max_depth,
                direction=direction,
            ):
                if node.id not in seen:
                    seen.add(node.id)
                    collected.append(node)
        self._current = collected
        return self

    def result(self) -> List[Node]:
        """Return the current list of matched/traversed nodes."""
        return list(self._current)

File: rust/src/query.rs
//! Graph traversal: BFS, DFS, shortest path and a fluent [`GraphQuery`].
//!
//! Traversals return node ids (matching the Phase-3 spec) and mirror the
//! Python `graphdb.query` semantics: the start node is excluded from BFS/DFS
//! results, `max_depth` bounds the walk, and `find_path` returns the shortest
//! path (including both endpoints).

use std::collections::{HashSet, VecDeque};

use crate::core::{Node, PropertyValue};
use crate::error::Result;
use crate::store::GraphStore;

/// Outgoing neighbours of `node_id`, optionally filtered by edge label.
fn neighbors(store: &GraphStore, node_id: &str, edge_label: Option<&str>) -> Vec<String> {
    store
        .edges_from(node_id)
        .into_iter()
        .filter(|e| edge_label.map(|l| e.label == l).unwrap_or(true))
        .map(|e| e.dst_id.clone())
        .collect()
}

/// Breadth-first traversal from `start_id` (excludes the start node).
pub fn bfs(
    store: &GraphStore,
    start_id: &str,
    edge_label: Option<&str>,
    max_depth: usize,
) -> Result<Vec<String>> {
    // Validate the start node exists.
    store.get_node(start_id)?;

    let mut visited: HashSet<String> = HashSet::new();
    visited.insert(start_id.to_string());
    let mut order: Vec<String> = Vec::new();
    let mut queue: VecDeque<(String, usize)> = VecDeque::new();
    queue.push_back((start_id.to_string(), 0));

    while let Some((current, depth)) = queue.pop_front() {
        if depth >= max_depth {
            continue;
        }
        for nb in neighbors(store, &current, edge_label) {
            if visited.insert(nb.clone()) {
                order.push(nb.clone());
                queue.push_back((nb, depth + 1));
            }
        }
    }
    Ok(order)
}

/// Depth-first traversal from `start_id` (excludes the start node).
pub fn dfs(
    store: &GraphStore,
    start_id: &str,
    edge_label: Option<&str>,
    max_depth: usize,
) -> Result<Vec<String>> {
    store.get_node(start_id)?;

    let mut visited: HashSet<String> = HashSet::new();
    visited.insert(start_id.to_string());
    let mut order: Vec<String> = Vec::new();

    fn visit(
        store: &GraphStore,
        node_id: &str,
        edge_label: Option<&str>,
        depth: usize,
        max_depth: usize,
        visited: &mut HashSet<String>,
        order: &mut Vec<String>,
    ) {
        if depth >= max_depth {
            return;
        }
        for nb in neighbors(store, node_id, edge_label) {
            if visited.insert(nb.clone()) {
                order.push(nb.clone());
                visit(store, &nb, edge_label, depth + 1, max_depth, visited, order);
            }
        }
    }

    visit(
        store,
        start_id,
        edge_label,
        0,
        max_depth,
        &mut visited,
        &mut order,
    );
    Ok(order)
}

/// Shortest path (list of node ids, endpoints included) from `src_id` to
/// `dst_id`, or `None` if unreachable.
pub fn find_path(store: &GraphStore, src_id: &str, dst_id: &str) -> Result<Option<Vec<String>>> {
    if store.get_node(src_id).is_err() || store.get_node(dst_id).is_err() {
        return Ok(None);
    }
    if src_id == dst_id {
        return Ok(Some(vec![src_id.to_string()]));
    }

    let mut visited: HashSet<String> = HashSet::new();
    visited.insert(src_id.to_string());
    let mut prev: std::collections::HashMap<String, String> = std::collections::HashMap::new();
    let mut queue: VecDeque<String> = VecDeque::new();
    queue.push_back(src_id.to_string());
    let mut found = false;

    while let Some(current) = queue.pop_front() {
        if current == dst_id {
            found = true;
            break;
        }
        for nb in neighbors(store, &current, None) {
            if visited.insert(nb.clone()) {
                prev.insert(nb.clone(), current.clone());
                queue.push_back(nb);
            }
        }
    }

    if !found && !prev.contains_key(dst_id) {
        return Ok(None);
    }

    // Reconstruct the path.
    let mut path: Vec<String> = vec![dst_id.to_string()];
    while path.last().map(|s| s.as_str()) != Some(src_id) {
        match prev.get(path.last().unwrap()) {
            Some(parent) => path.push(parent.clone()),
            None => return Ok(None),
        }
    }
    path.reverse();
    Ok(Some(path))
}

/// A small fluent query builder over a [`GraphStore`].
pub struct GraphQuery<'a> {
    store: &'a GraphStore,
    label_filter: Option<String>,
    prop_filters: Vec<(String, PropertyValue)>,
}

impl<'a> GraphQuery<'a> {
    pub fn new(store: &'a GraphStore) -> Self {
        GraphQuery {
            store,
            label_filter: None,
            prop_filters: Vec::new(),
        }
    }

    pub fn with_label(mut self, label: impl Into<String>) -> Self {
        self.label_filter = Some(label.into());
        self
    }

    pub fn with_property(mut self, key: impl Into<String>, value: PropertyValue) -> Self {
        self.prop_filters.push((key.into(), value));
        self
    }

    /// Execute the query, returning matching nodes.
    pub fn execute(self) -> Vec<&'a Node> {
        let label = self.label_filter.as_deref();
        // Start from the label-scoped set (or all nodes) then apply property
        // filters in memory.
        let mut nodes: Vec<&Node> = match label {
            Some(l) => self.store.nodes_by_label(l),
            None => self.store.all_nodes(),
        };
        for (key, value) in &self.prop_filters {
            nodes.retain(|n| n.properties.get(key).map(|pv| pv == value).unwrap_or(false));
        }
        nodes
    }
}
